### PR TITLE
[optim] ASGD fold state updates into functional and pass list of vars rather than states

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70953
* #70484
* #70483
* #70482
* #70481
* #70295
* #70231
* #70230
* #70229
* #69982
* #69981
* #69980
* __->__ #71335
* #71334
* #71333

Differential Revision: [D33767871](https://our.internmc.facebook.com/intern/diff/D33767871)

### CODE REVIEW

File: torch/optim/_multi_tensor/_functional.py [hunk 1]
```diff
@@ -162,28 +166,32 @@ def asgd(params: List[Tensor],
...
+            axs[i].copy_(params[i])
 
     # update eta and mu
-    for state in states:
-        state['eta'] = (lr /
-                        math.pow((1 + lambd * lr * state['step']), alpha))
-        state['mu'] = 1 / max(1, state['step'] - t0)
+    for i in range(len(mus)):
+        new_eta = torch.tensor(lr / math.pow((1 + lambd * lr * state_steps[i].item()), alpha))
+        etas[i].copy_(new_eta)
```
Comment (albanD) [inline torch/optim/_multi_tensor/_functional.py:192]: you can do `etas[i].fill_(XXX)` to avoid creating the tensor and the copy_

**Review by albanD — APPROVED** [review]
Sounds good!

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767871).

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767871).

**Pull request closed without merging**